Add AWS Custom Endpoint capability #70588

ampsingram opened the pull request:
Solves "Allow to override default AWS endpoint #70588"

Add several new properties to AWS CloudConfig to support custom endpoints.
Initialize/Parse on aws.go init() method which gets called when aws is loaded.
Allows overridden endpoints per servce and region. This allows functionality on air gapped networks.
This change is benign if services are not overridden in CloudConfig

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #70588 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Adds configuration for AWS endpoint fine control:
OverrideEndpoints bool Set to true to allow custom endpoints
ServiceDelimiter string Delimiter to use to separate overridden services (multiple services) Defaults to "&"
ServicenameDelimiter string Delimiter to use to separate servicename from its configuration parameters Defaults "|"
OverrideSeparator string Delimiter to use to separate region of occurrence, url and signing region for each override Defaults to ","
ServiceOverrides string example: s3|region1, https://s3.foo.bar, some signing_region & ec2|region2, https://ec2.foo.bar, signing_region

```

ampsingram committed:
Add AWS Custom Endpoint capability #70588

Solves "Allow to override default AWS endpoint #70588"

Add several new properties to AWS CloudConfig to support custom endpoints.
Initialize/Parse on aws.go init() method which gets called when aws is loaded.
Allows overridden endpoints per servce and region. This allows functionality on air gapped networks.
This change is benign if services are not overridden in CloudConfig

ampsingram commented:
Signed up Ampsight Inc as a corporate contributor, accepted the CLA

ampsingram commented:
Corporate CLA
I signed it

ingvagabund reviewed (commented):
Thank you for the PR. In overall, it looks good. I am just worried about the way of representing overriden services in `s3|region1, https://s3.foo.bar, sregion & s3|region2, https://s3.foo.bar, sregion` form. The line can get pretty long with many endpoints. Which makes it hard to read and configure.

ingvagabund commented:
gcfg allows to use multi-value variables. See https://godoc.org/gopkg.in/gcfg.v1#ReadStringInto, part `Example (Multivalue)`:
```
[global]
...
ServiceOverride=s3, us-east-1, https://s3.foo.bar, sregion
ServiceOverride=ec2, us-weat-1, https://s3.foo.bar, sregion
...
ServiceOverride=kms, us-east-2, https://s3.foo.bar, sregion
```

Reducing to only one delimiter, where the first component of each override 4-tuple needs to service name, then region, then endpoint, then signing region. Don't think it's necessary to have name on different level than the remaining regions and endpoint. The (name, region, endpoint, sregion) is easy to read as well.

ampsingram commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -569,6 +570,126 @@ type CloudConfig struct {
+			return defaultResolver.EndpointFor(service, region, optFns...)
+		}
+		return customResolverFn
+	} else {
```
the custom resolver is less efficient. If not being overridden, the default resolver is better.

ampsingram committed:
Updates for PR comments

ampsingram committed:
Improvement suggested by PR comment

Fail early makes the code more readable

ampsingram committed:
Fail early, helps readability

responding to a comment in the PR

ampsingram commented:
/check-cla

ampsingram commented:
https://github.com/kubernetes/kubernetes/pull/72245#issuecomment-455019465

But the map is signaled by the combined service name and region. Your example collides on s3|us-west-2 ; It is not a horrible idea, it has great merit, but that is a weakness that must be overcome.

micahhausler commented:
But when does cloud controller talk to more than one region? Can you think of a case where Kubernetes would actually need to call the services used (currently just EC2, ELB, ELBV2, KMS, and ASG) in more than one region?

ampsingram commented:
absolutely. S3 buckets can be in different regions. Other services, not so much, but still conceivable.

micahhausler commented:
Sure, but cloudprovider does not interact with S3, and as far as I’m aware, there are no plans to do so.

ampsingram commented:
in the interest of maximum flexibility, I would prefer to leave the services specified by both service name and region. I see no real downside to this and I think it provides better adherence to AWS's own rule that a service is specified by both a name and a region.

micahhausler commented:
I did think of an example for multi-region callout. [Kubelet calls ECR](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go) for different regions (depending on where an image is located). 

Ok, instead of keying off the section title, why not do this?

```go
package main

import (
	"fmt"

	"github.com/aws/aws-sdk-go/aws/endpoints"
	"gopkg.in/gcfg.v1"
)

type CloudConfig struct {
	Global struct {
		VPC string
		// ... other fields
	}
	ServiceOverride map[string]*struct {
		Region        string
		URL           string
		SigningRegion string
		Service       string
	}
}

func (c *CloudConfig) getResolver() endpoints.ResolverFunc {
	return func(service, region string, optFns ...func(*endpoints.Options)) (endpoints.ResolvedEndpoint, error) {
		for _, override := range c.ServiceOverride {
			if service == override.Service && region == override.Region {
				return endpoints.ResolvedEndpoint{
					URL:           override.URL,
					SigningRegion: override.SigningRegion,
				}, nil
			}
		}

		return endpoints.DefaultResolver().EndpointFor(service, region, optFns...)
	}
}

func main() {
	cfgStr := `
[Global]
vpc = vpc-abc1234567

[serviceoverride "ec2_us-west-2"]
URL = https://ec2.foo.bar
SigningRegion = custom-signing-region
Region = us-west-2
Service = ec2

[serviceoverride "ecr_us-west-2"]
URL = https://ecr.foo.bar
SigningRegion = custom-signing-region
Region = us-west-2
Service = ecr

[serviceoverride "ecr_us-east-1"]
URL = https://ecr.bar.foo
SigningRegion = custom-signing-region
Region = us-east-1
Service = ecr
`
	// change the CloudConfig type to have the "ServiceOverride" map
	cfg := &CloudConfig{}
	err := gcfg.ReadStringInto(cfg, cfgStr)
	if err != nil {
		panic(fmt.Errorf("Failed to parse gcfg data: %s", err))
	}

	cases := []struct {
		service string
		region  string
	}{
		{"ec2", "us-east-1"}, // not in config above
		{"ec2", "us-west-2"},
		{"ecr", "us-west-2"},
		{"ecr", "us-west-2"},
	}

	resolver := cfg.getResolver()

	for _, c := range cases {
		ep, err := resolver.EndpointFor(c.service, c.region)
		fmt.Printf("EndpointFor(%s, %s) got %+v, %+v\n", c.service, c.region, ep, err)
	}
}
```

which would output:

```
EndpointFor(ec2, us-east-1) got {URL:https://ec2.us-east-1.amazonaws.com SigningRegion:us-east-1 SigningName:ec2 SigningNameDerived:true SigningMethod:v4}, <nil>
EndpointFor(ec2, us-west-2) got {URL:https://ec2.foo.bar SigningRegion:custom-signing-region SigningName: SigningNameDerived:false SigningMethod:}, <nil>
EndpointFor(ecr, us-west-2) got {URL:https://ecr.foo.bar SigningRegion:custom-signing-region SigningName: SigningNameDerived:false SigningMethod:}, <nil>
EndpointFor(ecr, us-west-2) got {URL:https://ecr.foo.bar SigningRegion:custom-signing-region SigningName: SigningNameDerived:false SigningMethod:}, <nil>
```

ampsingram commented:
I do not like forcing the configuring user to construct map keys. This seems error prone. I still prefer the very simple ServiceOverrides that is currently implemented in this PR. I do not see why the map can not be constructed from simple parameters at init time. To represent this in the config seems a huge waste. What is the benefit?

ampsingram commented:
/check-cla

ampsingram commented:
/check-cla

ampsingram commented:
Does this PR introduce a user-facing change?:

Adds configuration for AWS endpoint fine control:

OverrideSeparator string Delimiter to use to separate service name, region of occurrence, url and signing region for each override Defaults to ","
ServiceOverrides []string 4 part Service name, region, url, signing region per line, Can occur multiple times
Example:
ServiceOverrides=s3, region1, https://s3.foo.bar, some signing_region
ServiceOverrides=ec2, region2, https://ec2.foo.bar, signing_region

micahhausler commented:
Personally, I know my team will be using this feature for the AWS cloud provider. By representing the service overrides as map keys, it is somewhat subjective, but to me it is more clear. Once you see one override, there isn't any cognitive load to remember order, and its not obvious looking at the comment documentation that the field can be repeated. For a human, how is constructing a map more error prone than a comma separated string? I'm not sure I see how representing a map is waste.

I'm not a big fan of the custom separator/parsing logic included in this PR. I know there are tests around it, but to me it seems pretty brittle. The order of fields matters, and any further fields (`SigningMethod`?) would need additional logic and tests to update.

I'm also wary of the mutable global `overrides` map. I get that its populated in `init()`, but I don't see any other mutable globals in `aws.go`, and its generally considered better practice to avoid mutable globals.

I'd rather see the parsing done by the library, and avoid using globals by putting a `getResolver()` method on `CloudConfig`. I think this is a great feature, but I'd just like to see the implementation be a little cleaner. I'll leave my comments here, and let someone else review and give their input.

ampsingram commented:
But ranging over keys with 2 comparisons is much less performant than a map lookup. The resolver will be used over and over. I prefer simplicity and performance. I guess others will have to chime in.

micahhausler commented:
Its true that a map lookup could be a little faster, but given the number of overrides is very likely to be less than 10 and certainly less than 100, iterating through a tight loop of small values only once at startup seems like an acceptable tradeoff.

ampsingram commented:
How about a compromise approach? We use the explicit config as you suggest, but I parse it on init. That would yield 2 advantages: one, map hash lookup and two, it would catch typos where a config was duplicated and the url changed, but not the region of occurrence. Thoughts?

ingvagabund commented:
With the map approach it is valid to do something like (duplicating map key, changing URL):

```
[serviceoverride "s3"]
Region= us-west-2
URL = https://s3.foo.bar
SigningRegion = custom-signing-region

[serviceoverride "s3"]
Region= us-west-2
URL = https://s3.foo.bar2222
SigningRegion = custom-signing-region

[serviceoverride "s3"]
Region= us-west-2
URL = https://s3.foo.bar2222333
SigningRegion = custom-signing-region
```

With the `gopkg.in/gcfg.v1` I have vendored in k8s, closer the duplicated map key to the end of the string is, the higher precedence it has . So the `URL` is set to `https://s3.foo.bar2222333` at the end. Though, I would not count on that being always true. Did not read the whole `gopkg.in/gcfg.v1` code/docs but I don't think the gcfg parser can be configured to fail in case of duplicated map keys.

I am against implementing another (yet simple) parser of the service overrides. Though, it's deterministic and we have control over dealing with duplicated keys.

ingvagabund commented:
> Adds configuration for AWS endpoint fine control:
ServiceDelimiter string Delimiter to use to separate overridden services (multiple services) Defaults to "&"
ServicenameDelimiter string Delimiter to use to separate servicename from its configuration parameters Defaults "|"
OverrideSeparator string Delimiter to use to separate region of occurrence, url and signing region for each override Defaults to ","
ServiceOverrides string example: s3|region1, https://s3.foo.bar, some signing_region & ec2|region2, https://ec2.foo.bar, signing_region

The PR description is no longer valid, can you update it to accordingly?

ingvagabund commented:
> I do not like forcing the configuring user to construct map keys. This seems error prone. I still prefer the very simple ServiceOverrides that is currently implemented in this PR. I do not see why the map can not be constructed from simple parameters at init time. To represent this in the config seems a huge waste. 

I am for it. Currently, there is only one delimiter of the service override value. It can be easily extended with other attributes in the future without breaking backward compatibility. E.g. going from `ServiceOverrides=s3, region1, https://s3.foo.bar, sregion` to `ServiceOverrides=s3, region1, https://s3.foo.bar, sregion, foo, bar`. Though, it is still important to keep the number of attributes low otherwise it will become harder to read.

Plus, `ServiceOverrides=s3, region1, https://s3.foo.bar, sregion` is easy to read.

micahhausler commented:
> How about a compromise approach? We use the explicit config as you suggest, but I parse it on init. That would yield 2 advantages: one, map hash lookup and two, it would catch typos where a config was duplicated and the url changed, but not the region of occurrence. Thoughts?

@ampsingram I could be OK with using the global map variable for now, where the `Services` interface change could be an optimization up for debate later. For this PR, I'd just like to confirm the file schema is using INI sections.

> @micahhausler +1 on that. I am against implementing another (yet simple) parser of the service overrides. Though, it's deterministic and we have control over dealing with duplicated keys.

> Plus, ServiceOverrides=s3, region1, https://s3.foo.bar, sregion is easy to read.

@ingvagabund I'm confused, are you leaning toward map keys or delimiter separated strings?

Since the cloud controller only talks to one region for each of the current 5 services, it should be pretty simple to visually debug each section

If this config were to be picked up by the kubelet for ECR endpoints in multiple regions, regions could be differentiated by changing the map key, since it could be discarded anyway.

```ini
[serviceoverride "ecr_us-west-2"]
URL = https://ecr.foo.bar
SigningRegion = custom-signing-region
Region = us-west-2
Service = ecr

[serviceoverride "ecr_us-east-1"]
URL = https://ecr.bar.bar
SigningRegion = custom-signing-region
Region = us-east-1
Service = ecr
```

ampsingram commented:
I have come around to Micah's thinking on the INI file structure. The map keys, to me, are nothing more than a service override identifier. They can be anything the user desires. In my tests, not yet committed, they were numbers, 1, 2, 3. The really nice thing about them is when I parse them and find errors or collisions, I can throw the error with the identifier. Should make the debug simpler. I can check for missing values, duplicated service and region entries, invalid URL, etc. I have no solution if the user duplicates identifiers. That is up to gcfg.

nckturner commented:
I also much prefer using the INI style for everything rather than overloading a single value.  @ampsingram if you can make that work in an acceptable way I think it would be easier to understand and result in simpler code.

micahhausler commented:
FWIW, its not a perfect test, but I was curious and ran a benchmark on a method looping over values versus a global map lookup. 

https://gist.github.com/micahhausler/fd2fefbcd9483a91a3c7ebc4dc4e59a0

```
$ go test -bench=. -count 5 -benchtime 2s  -benchmem
goos: darwin
goarch: amd64
pkg: github.com/micahhausler/configbench
BenchmarkGetResolverMethod-4   	 5000000	       435 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetResolverMethod-4   	10000000	       468 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetResolverMethod-4   	 5000000	       465 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetResolverMethod-4   	10000000	       426 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetResolverMethod-4   	10000000	       466 ns/op	       0 B/op	       0 allocs/op
BenchmarkGlobalMap-4           	 2000000	      1406 ns/op	     320 B/op	      18 allocs/op
BenchmarkGlobalMap-4           	 2000000	      1493 ns/op	     320 B/op	      18 allocs/op
BenchmarkGlobalMap-4           	 2000000	      1730 ns/op	     320 B/op	      18 allocs/op
BenchmarkGlobalMap-4           	 2000000	      1548 ns/op	     320 B/op	      18 allocs/op
BenchmarkGlobalMap-4           	 2000000	      1662 ns/op	     320 B/op	      18 allocs/op
PASS
ok  	github.com/micahhausler/configbench	45.293s
```

Its not a perfect match with your code, but I suspect the key construction for `fmt.Sprintf("%s__%s", strings.Trim()...)`  is what causes the global lookup to be 3x slower.

ampsingram committed:
Change to CloudConfig methods for validation and Resolver fn get
Move to a separate section for service overrides in INI, populate a struct for each override
update-bazel, update-gofmt and verify-spelling run

ampsingram committed:
Fix comment in CloudConfig s/Name/Service

ampsingram commented:
/ok-to-test

ampsingram committed:
gofmt changes

ampsingram commented:
/retest

ampsingram committed:
Fix golint findings

ampsingram commented:
/retest

ampsingram commented:
/test pull-kubernetes-e2e-kops-aws

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
+		Service       string
+		Region        string
+		URL           string
+		SigningRegion string
```
Can you add `SigningMethod` and pass that to [`endpoints.ResolvedEndpoint.SigningMethod`](https://docs.aws.amazon.com/sdk-for-go/api/aws/endpoints/#ResolvedEndpoint)?

ampsingram replied:
Can this be optional? In our testing, we did not set this and everything worked. Does it default to something sensible in us-east-1? That is where we were testing. In air gapped, maybe the default will be wrong?

micahhausler replied:
You're right that it will work without being set, as an empty value defaults to `v4`. This will help AWS with testing

ampsingram replied:
so use it as is, but do not validate it (check for empty)?

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
+	}
+	set := make(map[string]bool)
+	for onum, ovrd := range cfg.ServiceOverride {
+		name := strings.TrimSpace(ovrd.Service)
```
Is the TrimSpace necessary? Doesn't gcfg do this?

ampsingram replied:
Just added a test. No gcfg does not do this. That means I will have to add string trim to the comparisons when ranging over the map. Can I alter the map when validating? That seems preferable to trimming on every lookup.

micahhausler replied:
By my testing it looks like gcfg does trim extra space:

```go
package main

import (
	"fmt"

	"gopkg.in/gcfg.v1"
)

type CloudConfig struct {
	ServiceOverride map[string]*struct {
		Region        string
		URL           string
		SigningRegion string
		Service       string
		SigningMethod string
	}
}

func main() {
	cfgStr := `
[serviceoverride "ec2_us-west-2"]
URL=https://ec2.foo.bar
SigningRegion=custom-signing-region
Region=us-west-2
Service=ec2

[serviceoverride "ec2_us-east-1"]
URL = https://ec2.foo.bar
SigningRegion = custom-signing-region
Region = us-east-1
Service = ec2
`
	// change the CloudConfig type to have the "ServiceOverride" map
	cfg := &CloudConfig{}
	err := gcfg.ReadStringInto(cfg, cfgStr)
	if err != nil {
		panic(fmt.Errorf("Failed to parse gcfg data: %s", err))
	}
	for _, override := range cfg.ServiceOverride {
		fmt.Printf("Region: '%s'\n", override.Region)
		fmt.Printf("URL: '%s'\n", override.URL)
		fmt.Printf("SigningRegion: '%s'\n", override.SigningRegion)
		fmt.Printf("Service: '%s'\n", override.Service)
	}
}
```

spits out:

```
Region: 'us-west-2'
URL: 'https://ec2.foo.bar'
SigningRegion: 'custom-signing-region'
Service: 'ec2'
SigningMethod: ''
Region: 'us-east-1'
URL: 'https://ec2.foo.bar'
SigningRegion: 'custom-signing-region'
Service: 'ec2'
```

ampsingram replied:
to add spaces in INI, you have to quote the string

ampsingram replied:
`{
			"Active Overrides",
			strings.NewReader(`
                [Global]

               [ServiceOverride "1"]
                Service = "s3      "
                Region = sregion
                URL = https://s3.foo.bar
                SigningRegion = sregion
                SigningMethod = sign
                `),
			nil,
			false, true,
			[]ServiceDescriptor{{name: "s3", region: "sregion"}},
		},
This failed without trims`

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
+	return nil
+}
+
+func (cfg *CloudConfig) getResolver() func(service, region string,
```
Can you change this signature to 
```go
func (cfg *CloudConfig) getResolver() endpoints.ResolverFunc {
}
```

ampsingram replied:
yes

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
+		return defaultResolver.EndpointFor(service, region, optFns...)
+	}
+	if len(cfg.ServiceOverride) == 0 {
+		return defaultResolverFn
```
How about 
```go
response := endpoints.DefaultResolver().EndpointFor(service, region, optFns...)
if len(cfg.ServiceOverride) == 0 {
    return response
}
```

ampsingram replied:
That is the complexity of functions returning functions.  endpoints.DefaultResolver().EndpointFor(service, region, optFns...) is what is executed, we have to wrap it to make it a func for return.

micahhausler replied:
ah you're right

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
+	if len(cfg.ServiceOverride) == 0 {
+		return defaultResolverFn
+	}
+	customResolverFn := func(service, region string,
```
Nit: rather than assigning to the variable `customResolverFn`, why not just `return func(...`?

ampsingram replied:
reduces comprehension a bit, but ok.

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,102 @@ type CloudConfig struct {
 
+// Interface to make the CloudConfig immutable for awsSDKProvider
+type awsCloudConfigProvider interface {
+	getResolver() func(string, string, ...func(*endpoints.Options)) (endpoints.ResolvedEndpoint, error)
```
Can you update this signature to 
```go
type awsCloudConfigProvider interface{
    getResolver() endpoints.ResolverFunc 
}
```

ampsingram replied:
yes

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -651,7 +731,8 @@ func (p *awsSDKProvider) Compute(regionName string) (EC2, error) {
 	}
-	awsConfig = awsConfig.WithCredentialsChainVerboseErrors(true)
+	awsConfig = awsConfig.WithCredentialsChainVerboseErrors(true).
+		WithEndpointResolver(endpoints.ResolverFunc(p.cfg.getResolver()))
```
You don't need the `endpoints.ResolverFunc()` wrapper, can you just do 

```go
awsConfig =  awsConfig.WithCredentialsChainVerboseErrors(true).
        WithEndpointResolver(p.cfg.getResolver())
```

ampsingram replied:
yes, not needed since I changed the getResolver signature. That cast came from the AWS example.

micahhausler reviewed (changes requested):
This is looking great. Just a few changes

ampsingram committed:
Add SigningMethod, make it optional, copy in verbatim

An empty SigningMethod will default properly.

ampsingram committed:
format changes

ampsingram commented:
changes made, the one test failure appears to be resource related.

ampsingram commented:
/test pull-kubernetes-kubemark-e2e-gce-big

ampsingram commented:
/retest

micahhausler commented on pkg/cloudprovider/providers/aws/aws.go:
```diff
@@ -570,23 +571,113 @@ type CloudConfig struct {
+		Region        string
+		URL           string
+		SigningRegion string
+		SigningMethod string
```
One more request before this goes, in, could you add `SigningName string` field?  It doesn't need to be a documented field, but it will really help AWS out with testing. An [empty string is value is fine](https://github.com/aws/aws-sdk-go/blob/173fcf0a78d6bc6c1668a416cdef0032c2a0efab/private/model/api/api.go#L318-L322), the SDK will just use the correct default.

ampsingram replied:
done

micahhausler commented:
/lgtm

ampsingram committed:
Add SigningName as optional parameter

Makes AWS testing simpler

ampsingram commented:
/test pull-kubernetes-integration

ampsingram commented:
/test pull-kubernetes-integration

ampsingram commented:
/test pull-kubernetes-e2e-kops-aws

micahhausler commented:
/test pull-kubernetes-e2e-kops-aws
/lgtm 
/approve

ampsingram commented:
/test pull-kubernetes-e2e-kops-aws
